### PR TITLE
Codemod

### PR DESCRIPTION
**Pull request by GerardasB**

### CODE REVIEW

File: tools/codemod/src/v4.0.0/__tests__/frontstage-to-config.test.ts [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+          }],
+
+          end: [{
+            id: 3,
+          }],
+        },
+      },
+    })
+    `,
+    "transforms rightPanel correctly1"
```
Comment (GerardasB) [inline tools/codemod/src/v4.0.0/__tests__/frontstage-to-config.test.ts:60]: @AntanasBuk test names should be descriptive, especially when there are different variations of the same test/functionality.

File: tools/codemod/src/v4.0.0/Utils/jsxElementAttributeHandles.ts [hunk 2]
```diff
@@ -0,0 +1,388 @@
+/*---------------------------------------------------------------------------------------------
```
Comment (GerardasB) [inline tools/codemod/src/v4.0.0/Utils/jsxElementAttributeHandles.ts:1]: Move to 'src/utils'.

File: tools/codemod/src/utils/testUtils.ts [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+function createCollectionTransform(transformer: CollectionTransformer): Transform {
+  return (file, api, options) => {
+    const j = api.jscodeshift;
+    const root = j(file.source);
+    transformer(j, root);
+    return root.toSource(options.printOptions);
+  };
+}
+
+// export function createInlineTransform(transform: Transform): Transform {
```
Comment (GerardasB) [inline tools/codemod/src/utils/testUtils.ts:66]: Dead code - remove.

File: tools/codemod/src/utils/typeGuards.ts [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+* See LICENSE.md in the project root for license terms and full copyright notice.
+*--------------------------------------------------------------------------------------------*/
+import type {
+  ArrayExpression, ASTNode, CallExpression, Identifier, ImportSpecifier, JSCodeshift, JSXAttribute, JSXElement, JSXEmptyExpression,
+  JSXExpressionContainer, JSXIdentifier, Literal, MemberExpression, ObjectExpression,
+} from "jscodeshift";
+
+type Source = Parameters<JSCodeshift>[0];
+
+export function isJSXIdentifierType(j: JSCodeshift, source: Source): source is JSXIdentifier {
```
Comment (GerardasB) [inline tools/codemod/src/utils/typeGuards.ts:12]: Only type guards that take `ASTNode` as source are correct, @AntanasBuk is working on refactoring this.

File: tools/codemod/src/utils/Extensions.ts [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+      const sourceDeclarations = this.findImportDeclarations(source.module);
+      const sourceSpecifiers = sourceDeclarations.findSpecifiers(source.specifier);
+
+      const sourceLocalSpecifierName = sourceSpecifiers.getLocalName() || source.specifier;
+      const sourceExpr = updateExpressionSpecifier(source.expr, sourceLocalSpecifierName);
+
+      retainFirstComment(j, this, () => {
+        let updateSpecifiers = false;
+        if (source.isIdentifier) {
+          const rootIdentifiers = this.find(j.Identifier)
```
Comment (GerardasB) [inline tools/codemod/src/utils/Extensions.ts:59]: Use `findRootIdentifiers`.

File: tools/codemod/src/utils/CallExpression.ts [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+  };
+  j.registerMethods(globalMethods);
+  j.registerMethods(methods, j.CallExpression);
+}
+
+export function useCallExpression(j: JSCodeshift) {
+  usePlugin(j, callExpressionPlugin);
+}
+
+export function objectExpressionFilter(j: JSCodeshift): ArgumentsFilter {
```
Comment (GerardasB) [inline tools/codemod/src/utils/CallExpression.ts:62]: Remove, just inline that..

File: tools/codemod/LICENSE.md [hunk 7]
```diff
@@ -0,0 +1,9 @@
+# MIT License
+
+Copyright © 2017-2023 Bentley Systems, Incorporated. All rights reserved.
```
Comment (aruniverse) [inline tools/codemod/LICENSE.md:3]:
```suggestion
Copyright © 2023 Bentley Systems, Incorporated. All rights reserved.
```

**Review by aruniverse — COMMENTED** [review]
i didnt review the actual codemod files
i assume/hope they all have the copyright header

**Comment by aruniverse** [comment]
@skirby1996 mind taking a pass at this new pkg? and the publishing pipeline changes associated with it?

File: common/config/azure-pipelines/jobs/version-bump-individual.yaml [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+        inputs:
+          versionSpec: 16.x
+          checkLatest: true
+
+      - bash: |
+          git config --local user.email imodeljs-admin@users.noreply.github.com
+          git config --local user.name imodeljs-admin
+        displayName: Setup Git
+
+      - bash: "node common/scripts/install-run-rush version --bump --version-policy ${{ parameters.VersionPolicy }}"
```
Comment (skirby1996) [inline common/config/azure-pipelines/jobs/version-bump-individual.yaml:37]: Does this correctly bump the version or prompt for any additional input? I'm not sure how rush version --bump works with an "individualVersion" policy.

Comment (GerardasB) [reply common/config/azure-pipelines/jobs/version-bump-individual.yaml:37]: Yes it bumps the versions based on previously gathered `rush change` inputs, i.e. each `rush change` asks to input both the change message and version increment (path/minor/major/).

**Comment by GerardasB** [comment]
> i didnt review the actual codemod files i assume/hope they all have the copyright header

Yes, we still run the copyright-linter...

**Review by skirby1996 — APPROVED** [review]

**Review by raplemie — APPROVED** [review]
I did not review all the tests and text fixture, lot to cover here...

Do we have a minimum version on which these codemods are expected to work? 
If someone use that on a codebase that is 2.19, what would happen ?

File: tools/codemod/.npmignore [hunk 9]
```diff
@@ -0,0 +1,9 @@
+# start off ignoring everything, and then add back only the files we want
+*
+
+!*.md
+
+!src/**/*.ts
```
[the remark this answers is not in the document]
Comment (GerardasB) [reply tools/codemod/.npmignore:6]: It doesn't get ignored by npm since it's a `bin` script.

**Comment by GerardasB** [comment]
> ommon/config/rush/pnpm-lock.yam

We expect these to be run by applications already upgraded to 3.x.
Running from 2.x is not tested, but shouldn't be destructive, would only update less code depending on how different 2.x and 3.x versions are.

**Pull request merged**